File: backend/app/models/song.py
"""Song data models."""
from pydantic import BaseModel
from typing import Optional, List
from datetime import datetime
from enum import Enum


class Difficulty(str, Enum):
    """Song difficulty level."""
    BEGINNER = "beginner"
    INTERMEDIATE = "intermediate"
    ADVANCED = "advanced"


class Source(str, Enum):
    """Song source."""
    LIBRARY = "library"
    USER_UPLOAD = "user_upload"
    YOUTUBE = "youtube"
    SPOTIFY = "spotify"


class TranscriptionStatus(str, Enum):
    """Transcription status."""
    PENDING = "pending"
    PROCESSING = "processing"
    COMPLETED = "completed"
    FAILED = "failed"


class SongMetadata(BaseModel):
    """Song metadata."""
    spotify_id: Optional[str] = None
    youtube_id: Optional[str] = None
    musicbrainz_id: Optional[str] = None
    album_art: Optional[str] = None
    source: Source = Source.LIBRARY


class AudioFiles(BaseModel):
    """Audio file paths."""
    original: Optional[str] = None
    instrumental: Optional[str] = None
    separated: Optional[dict] = None  # {vocals, drums, bass, other}


class TranscriptionInfo(BaseModel):
    """Transcription information."""
    status: TranscriptionStatus = TranscriptionStatus.PENDING
    processed_at: Optional[datetime] = None
    instruments: List[str] = []


class Song(BaseModel):
    """Song model."""
    song_id: str
    title: str
    artist: str
    album: Optional[str] = None
    duration: int  # seconds
    genre: Optional[str] = None
    difficulty: Difficulty = Difficulty.BEGINNER
    original_key: Optional[str] = None
    tempo: Optional[int] = None  # BPM
    time_signature: Optional[str] = None
    metadata: SongMetadata = SongMetadata()
    audio_files: AudioFiles = AudioFiles()
    transcription: TranscriptionInfo = TranscriptionInfo()
    created_by: Optional[str] = None
    created_at: datetime
    is_public: bool = True
    download_count: int = 0
    favorite_count: int = 0


class SongCreate(BaseModel):
    """Song creation request."""
    title: str
    artist: str
    album: Optional[str] = None
    duration: int
    genre: Optional[str] = None
    difficulty: Difficulty = Difficulty.BEGINNER


class SongResponse(BaseModel):
    """Song response."""
    song_id: str
    title: str
    artist: str
    album: Optional[str] = None
    duration: int
    genre: Optional[str] = None
    difficulty: Difficulty
    original_key: Optional[str] = None
    tempo: Optional[int] = None
    time_signature: Optional[str] = None
    metadata: SongMetadata
    available_instruments: List[str]
    is_favorite: bool = False
    is_downloaded: bool = False


class SongSearchResult(BaseModel):
    """Song search result."""
    song_id: str
    title: str
    artist: str
    album_art: Optional[str] = None
    duration: int
    source: Source

File: frontend/lib/models/song.dart
class Song {
  final String songId;
  final String title;
  final String artist;
  final String? album;
  final int duration; // seconds
  final String? genre;
  final String? difficulty;
  final String? originalKey;
  final int? tempo;
  final String? timeSignature;
  final SongMetadata metadata;
  final List<String> availableInstruments;
  final bool isFavorite;
  final bool isDownloaded;

  Song({
    required this.songId,
    required this.title,
    required this.artist,
    this.album,
    required this.duration,
    this.genre,
    this.difficulty,
    this.originalKey,
    this.tempo,
    this.timeSignature,
    required this.metadata,
    required this.availableInstruments,
    this.isFavorite = false,
    this.isDownloaded = false,
  });

  factory Song.fromJson(Map<String, dynamic> json) {
    return Song(
      songId: json['songId'] as String,
      title: json['title'] as String,
      artist: json['artist'] as String,
      album: json['album'] as String?,
      duration: json['duration'] as int,
      genre: json['genre'] as String?,
      difficulty: json['difficulty'] as String?,
      originalKey: json['originalKey'] as String?,
      tempo: json['tempo'] as int?,
      timeSignature: json['timeSignature'] as String?,
      metadata: SongMetadata.fromJson(json['metadata'] as Map<String, dynamic>),
      availableInstruments: (json['availableInstruments'] as List<dynamic>?)
              ?.map((e) => e as String)
              .toList() ??
          [],
      isFavorite: json['isFavorite'] as bool? ?? false,
      isDownloaded: json['isDownloaded'] as bool? ?? false,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'songId': songId,
      'title': title,
      'artist': artist,
      'album': album,
      'duration': duration,
      'genre': genre,
      'difficulty': difficulty,
      'originalKey': originalKey,
      'tempo': tempo,
      'timeSignature': timeSignature,
      'metadata': metadata.toJson(),
      'availableInstruments': availableInstruments,
      'isFavorite': isFavorite,
      'isDownloaded': isDownloaded,
    };
  }

  String get formattedDuration {
    final minutes = duration ~/ 60;
    final seconds = duration % 60;
    return '$minutes:${seconds.toString().padLeft(2, '0')}';
  }

  Song copyWith({
    bool? isFavorite,
    bool? isDownloaded,
  }) {
    return Song(
      songId: songId,
      title: title,
      artist: artist,
      album: album,
      duration: duration,
      genre: genre,
      difficulty: difficulty,
      originalKey: originalKey,
      tempo: tempo,
      timeSignature: timeSignature,
      metadata: metadata,
      availableInstruments: availableInstruments,
      isFavorite: isFavorite ?? this.isFavorite,
      isDownloaded: isDownloaded ?? this.isDownloaded,
    );
  }
}

class SongMetadata {
  final String? spotifyId;
  final String? youtubeId;
  final String? musicbrainzId;
  final String? albumArt;
  final String source; // library, user_upload, youtube, spotify

  SongMetadata({
    this.spotifyId,
    this.youtubeId,
    this.musicbrainzId,
    this.albumArt,
    required this.source,
  });

  factory SongMetadata.fromJson(Map<String, dynamic> json) {
    return SongMetadata(
      spotifyId: json['spotifyId'] as String?,
      youtubeId: json['youtubeId'] as String?,
      musicbrainzId: json['musicbrainzId'] as String?,
      albumArt: json['albumArt'] as String?,
      source: json['source'] as String,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'spotifyId': spotifyId,
      'youtubeId': youtubeId,
      'musicbrainzId': musicbrainzId,
      'albumArt': albumArt,
      'source': source,
    };
  }
}
